crates/merkle-inclusion/tests/gadget.rs

```diff
@@ -0,0 +1,370 @@
+            }
+        }
+    };
+}
+
+// Generate tests for Sha3 and Keccak with lowercase suffixes
+create_hash_tests!(sha3, Sha3);
+create_hash_tests!(keccak, Keccak);
+
+/**************************************************************
+ * Utilities
+ **************************************************************/
+
+pub fn bytes_to_bitvec(bytes: &[u8]) -> Vec<Boolean> {
+    let bits = BitVec::<Lsb0, u8>::from_slice(bytes);
+    let bits: Vec<Boolean> = bits.iter().map(|b| Boolean::constant(*b)).collect();
+    bits
+}
+
+pub fn bits_to_bytevec(bits: &[Boolean]) -> Vec<u8> {
+    let result: Vec<bool> = bits.iter().map(|b| b.get_value().unwrap()).collect();
+    let mut bv = BitVec::<Lsb0, u8>::new();
+    for bit in result {
+        bv.push(bit);
+    }
+    bv.as_slice().to_vec()
+}
+
+pub struct SimpleMerkleTree {
+    root: Vec<u8>,
+    leaf_key_vec: Vec<(Vec<u8>, Vec<Boolean>)>,
+}
+
+impl SimpleMerkleTree {
+    pub fn get_leaf_key(&self, index: usize) -> &[Boolean] {
+        &self.leaf_key_vec.get(index).unwrap().1
+    }
+
+    pub fn get_leaf_hash(&self, index: usize) -> &[u8] {
+        &self.leaf_key_vec.get(index).unwrap().0
+    }
+
+    pub fn root(&self) -> &[u8] {
+        &self.root
+    }
+}
+
+/// Constructs a fixed-depth binary Merkle tree with predefined leaf values.
+///
+/// This function creates a Merkle tree with a depth of 4 (including the root).
+/// The leaves of the tree are the SHA3 hashes of the predefined values: "a", "b", "c", "d", "e", "f", "g", "h".
+/// It computes the hash and path key for each leaf and intermediary node.
+///
+/// # Returns
+/// A tuple containing:
+/// - The root hash of the Merkle tree as the first element.
+/// - A vector of tuples, each containing:
+///   - The hash of a leaf or intermediary node.
+///   - The path key, a `Vec<Boolean>`, representing the path from the root to the node.
+///
+/// # Path Key Representation
+/// The path key is a vector of `Boolean` values, where each value indicates a direction at a node:
+/// - `Boolean::constant(false)` for a left turn.
+/// - `Boolean::constant(true)` for a right turn.
+/// The path keys are 256 bits long, padded with `false` values.
+///
+/// # Tree Structure
+/// The tree structure is as follows (indexes in the vector are shown in parentheses):
+/// ```plaintext
+///         root
+///        /    \
+///      (12)   (13)
+///      /  \    /  \
+///    (8) (9) (10) (11)
+///    / \  / \  / \  / \
+///  (0)(1)(2)(3)(4)(5)(6)(7)
+///   a  b  c  d  e  f  g  h
+/// ```
+/// In this structure:
+/// - Leaf nodes ("a" to "h") are at indexes 0 to 7.
+/// - Intermediary nodes are at indexes 8 to 13.
+/// - The root node is at index 14.
+///
+/// # Example
+/// ```
+/// let (root_hash, leaf_hashes_and_keys) = construct_merkle_tree();
+/// for (hash, key) in leaf_hashes_and_keys.iter() {
+///     // Process each leaf and intermediary node's hash and key...
+/// }
+/// ```
+pub fn construct_merkle_tree<D: Digest>() -> SimpleMerkleTree {
```

Comment:
It's not needed for this iteration, but we'll probably want to open an issue to use proptest to generate this sort of test instance.

Reply:
@huitseeker I just added it to have a clean tree to use for testing. Would you rather have me removing this implementation and implement a more straight forward construction of the leafs / proofs?

Reply:
No, it's ok. I'm just mentioning specifically that `proptest_recurse` can generate random tree-structured data:
https://docs.rs/proptest-recurse/latest/proptest_recurse/
This is probably something we'll want, but in the future.